sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -59,7 +59,10 @@ trait TimeZoneAwareExpression extends Expression {
   /** Returns a copy of this expression with the specified timeZoneId. */
   def withTimeZone(timeZoneId: String): TimeZoneAwareExpression
 
-  @transient lazy val zoneId: ZoneId = DateTimeUtils.getZoneId(timeZoneId.get)
```

Comment:
It's guaranteed to have `timeZoneId` defined by `resolved` above. How did you face the error case?

Reply:
Using the expression outside the context of a DataFrame gives the error. For example, the functions hour, date_format on a spark shell will give the error.

```
scala> hour(current_timestamp).expr.eval()
java.util.NoSuchElementException: None.get
  at scala.None$.get(Option.scala:529)
  at scala.None$.get(Option.scala:527)
  at org.apache.spark.sql.catalyst.expressions.TimeZoneAwareExpression.zoneId(datetimeExpressions.scala:53)
  at org.apache.spark.sql.catalyst.expressions.TimeZoneAwareExpression.zoneId$(datetimeExpressions.scala:53)
  at org.apache.spark.sql.catalyst.expressions.Hour.zoneId$lzycompute(datetimeExpressions.scala:321)
  at org.apache.spark.sql.catalyst.expressions.Hour.zoneId(datetimeExpressions.scala:321)
  at org.apache.spark.sql.catalyst.expressions.GetTimeField.nullSafeEval(datetimeExpressions.scala:302)
  at org.apache.spark.sql.catalyst.expressions.GetTimeField.nullSafeEval$(datetimeExpressions.scala:301)
  at org.apache.spark.sql.catalyst.expressions.Hour.nullSafeEval(datetimeExpressions.scala:321)
  at org.apache.spark.sql.catalyst.expressions.UnaryExpression.eval(Expression.scala:476)
  ... 47 elided

scala> date_format(current_timestamp, "dd").expr.eval()
java.util.NoSuchElementException: None.get
  at scala.None$.get(Option.scala:529)
  at scala.None$.get(Option.scala:527)
  at org.apache.spark.sql.catalyst.expressions.TimeZoneAwareExpression.zoneId(datetimeExpressions.scala:53)
  at org.apache.spark.sql.catalyst.expressions.TimeZoneAwareExpression.zoneId$(datetimeExpressions.scala:53)
  at org.apache.spark.sql.catalyst.expressions.DateFormatClass.zoneId$lzycompute(datetimeExpressions.scala:772)
  at org.apache.spark.sql.catalyst.expressions.DateFormatClass.zoneId(datetimeExpressions.scala:772)
  at org.apache.spark.sql.catalyst.expressions.TimestampFormatterHelper.getFormatter(datetimeExpressions.scala:70)
  at org.apache.spark.sql.catalyst.expressions.TimestampFormatterHelper.getFormatter$(datetimeExpressions.scala:67)
  at org.apache.spark.sql.catalyst.expressions.DateFormatClass.getFormatter(datetimeExpressions.scala:772)
  at org.apache.spark.sql.catalyst.expressions.TimestampFormatterHelper.$anonfun$formatterOption$1(datetimeExpressions.scala:64)
  at scala.Option.map(Option.scala:230)
```

Reply:
There expressions are not supposed to be used standalone, and these are supposed to be internal (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/package.scala#L21-L22).

Reply:
I agree that using Expressions are not supposed to be used standalone, but the error is not due the usage of expressions, it's during the usage of Column. It's a common practise that people use Column type as the argument to a transformation function like the usage of `lit("..")`. Later in the transformation pipeline if we want to get a string value from the column type, we should be able to get it back. 

In our use case, we pass the value of a date as Column type at the beginning of the transformation pipeline to do the filtering on different DFs and at the end of the transformation, we use the date value for partitioning the results. Currently, we don't have better choice other than passing the date value as string type and convert it to column type using `lit(..)` during filtering which is not so elegant/ consistent with the rest of the pipeline.

Reply:
The expression is only resolved when the time zone has been set:
https://github.com/apache/spark/blob/6d4223071f4cf475ce5c285e6b3a96cd93f5b166/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala#L46-L48

Reply:
I agree, but it will get resolved only when used in the context of a DataFrame/Dataset. Using a Column type outside, for example passing it as an argument to a function to do the filtering will show a confusing error `java.util.NoSuchElementException: None.get`.

Reply:
If you are ready to use internal stuff, just set the time zone manually:
```scala
scala> hour(current_timestamp).expr.asInstanceOf[TimeZoneAwareExpression].withTimeZone("UTC").eval()
res6: Any = 20

scala> hour(current_timestamp).expr.asInstanceOf[TimeZoneAwareExpression].withTimeZone("Europe/Moscow").eval()
res7: Any = 23
```